cpp/src/arrow/dataset/partition.h

```diff
@@ -63,13 +64,18 @@ struct ARROW_DS_EXPORT PartitionPathFormat {
 /// Paths are consumed from left to right. Paths must be relative to
 /// the root of a partition; path prefixes must be removed before passing
 /// the path to a partitioning for parsing.
-class ARROW_DS_EXPORT Partitioning {
+class ARROW_DS_EXPORT Partitioning : public util::EqualityComparable<Partitioning> {
  public:
   virtual ~Partitioning() = default;
 
   /// \brief The name identifying the kind of partitioning
   virtual std::string type_name() const = 0;
 
+  /// \brief determines if partiioning is exactly equal
+  virtual bool Equals(const Partitioning& other, bool check_metadata = false) const {
```

Comment:
We shouldn't need a `check_metadata` flag here.  This method here should pass `check_metadata=false` in always.

If, by some chance, a future partitioning decides to depend on the schema metadata (which should be possible) then it should override this method and pass `check_metadata=true`.